csrc/preseg_passes/allocation_order_inference.cpp

```diff
@@ -166,69 +221,26 @@ void AllocationOrderInferencer::handle(BroadcastOp* op) {
   alloc_order_map_[out] = permutation.value();
 }
 
-// BinaryOp propagation tries to merge the allocation order of both inputs
-//
-//   1. when there's only one operand has a recorded allocation order, it
-//   forwards that. This could happen when: we have inputs without recorded
-//   allocation order, or one of the operands being a scalar;
-//   2. When both tensor have recorded allocation order. The one tensor with
-//   more non-broadcast iterdomain will be dominating the output allocation
-//   order. The motivation behind it to avoid breaking allocation order
-//   propagation from binary operation against unsqueezed vector tensors.
-//
-//   In the event of a tie, we'll just propagate the allocation order of lhs.
 void AllocationOrderInferencer::handle(BinaryOp* op) {
   auto* out = dynamic_cast<TensorView*>(op->out());
   if (out == nullptr) {
     return;
   }
-  auto* lhs = dynamic_cast<TensorView*>(op->lhs());
-  auto* rhs = dynamic_cast<TensorView*>(op->rhs());
-  if (lhs == nullptr) {
-    // propagate rhs when lhs is not a tensor
-    // Note: rhs could also be non-tensor, which we'll just skip setting
-    // allocation order for output and return
-    propagateAllocationOrder(rhs, out);
-    return;
-  }
-  if (rhs == nullptr) {
-    // propagate lhs when rhs is not a tensor
-    propagateAllocationOrder(lhs, out);
-    return;
-  }
-
-  // both operands are tensors
-  auto lhs_iter = alloc_order_map_.find(lhs);
-  auto rhs_iter = alloc_order_map_.find(rhs);
-
-  if (rhs_iter == alloc_order_map_.end()) {
-    propagateAllocationOrder(lhs, out);
-    return;
-  }
-  if (lhs_iter == alloc_order_map_.end()) {
-    propagateAllocationOrder(rhs, out);
-    return;
-  }
+  propagateAllocationOrder(resolveAllocationOrder(op->inputs()), out);
+}
 
-  // we have allocation order recorded for both tensor
-  // if both allocation order agree, we just propagate it
-  if (lhs_iter->second == rhs_iter->second) {
-    alloc_order_map_[out] = lhs_iter->second;
+void AllocationOrderInferencer::handle(TernaryOp* op) {
+  auto* out = dynamic_cast<TensorView*>(op->out());
+  if (out == nullptr) {
     return;
   }
-  auto countNonBroadcastID = [](const TensorView* tv) {
-    return std::count_if(
-        tv->getMaybeRFactorDomain().begin(),
-        tv->getMaybeRFactorDomain().end(),
-        [&](auto ptr_id) { return !ptr_id->isBroadcast(); });
-  };
+  propagateAllocationOrder(resolveAllocationOrder(op->inputs()), out);
+}
 
-  // otherwise, we propagate the one with more non-broadcast iterdomains.
-  if (countNonBroadcastID(lhs) >= countNonBroadcastID(rhs)) {
-    alloc_order_map_[out] = lhs_iter->second;
-  } else {
-    alloc_order_map_[out] = rhs_iter->second;
-  }
+void AllocationOrderInferencer::handle(PadOp* op) {
+  auto* out = dynamic_cast<TensorView*>(op->out());
+  auto* in = dynamic_cast<TensorView*>(op->in());
+  propagateAllocationOrder(in, out);
```

Comment:
pad is trivial and *almost* identical to unary. so I didn't bother adding a test.